Mid-Summer PR

CalciumKing opened the pull request:
Second draft of the common stock inventory management project. Fixed some bugs and changed backend functionality with last PR's comments in mind. Some GUI and core functionality changes were made like adding a login page, QR generation/exportation, and backend functionality changes, efficiencies, and optimizations. Google Sheets API has been integrated and works completely for any aspects that require viewing or editing of the Google Sheet. Global keyboard shortcuts have made the user experience better.

## Description
Changed several `.ui` files to be more efficient by resetting default values. Some `.ui` files have also been added to for styling, new pages, login functionality, etc. `Enum` objects were added to hold page data that is referenced by different parts of the program to make the process of adding pages and viewing certain page data easier, this was a topic of discussion in the last PR. Users can now log into the application using either their QR code or with their username, no password is required because the current system does not require a login of any kind, so a username for logging purposes would be sufficient. Lots of icons were added simply to make the application look better and better styling, all are located in `assets/icons`. `README.md` has been updated appropriately to reflect the current state of this project. Several files like `README.md`, `meta.yaml`, and `requirements.txt` have been updated with the projects current build requirements. Assets have been added to `resources.qrc` for cross platform compatibility, however it is not currently used in any `.ui` files because of an issue my device was having with parsing the file to a `.py` file. Global keyboard shortcuts are now available on all pages, however some are restricted on the login page. The user can also locate and view all the shortcuts on the menu bar in the top left corner of the screen.

## Motivation and Context
This PR updates the previous one with more updated, styled, and efficient code. Some application pages have an updated `.ui` file to update the functionality of the page. Backend changes make the process of creating a new page and managing all page data easier, which was a concern with the last PR. Login page functionality has also significantly improved by creating a valid login page that works as intended and will not crash the program or cause issues with bypassing the login page. Table functionality has also been fixed and all tables display the same information, all in the same order as the database, which was a significant bug in the last PR.

## How Has This Been Tested?
Basic bug testing by myself has happened, and multiple bugs have been fixed. A major fix was the login page and preventing the application from crashing if the camera is being used from both the login page and the item scanner simultaneously. As of the most recent push, there are no known bugs, however further and more legitimate testing will be conducted within in the next few weeks.

## Where Has This Been Documented?
Primary documentation is this PR. Most files, classes, methods. etc. have new or updated docstrings attached. Each commit has individual details explaining what changes were made in that commit and why.

<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/138fd17f-4f20-48aa-80f5-5788dfff49b9" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/f610f36a-19a0-449c-a5a8-4de78ad4b4eb" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/a4de7fc1-85ff-4ea5-94b0-ccb5b2ba7db2" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/fadb4697-40c7-464f-be79-14f1236e1c0a" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/a390c38a-120c-48db-bfaa-3eb4f24edc29" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/2bb287d8-03c8-4531-b61b-9e292807fb69" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/2ffa2ffb-a651-46fb-aba2-01081ad9c436" />
<img width="1920" height="1030" alt="image" src="https://github.com/user-attachments/assets/9deb480a-fd8d-4546-9ed5-804d71152277" />


## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [ ] Test suite passes locally
- [ ] Test suite passes on GitHub Actions
- [ ] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

CalciumKing committed:
Added username logging in informational logging messages, fixed logging and print messages.

CalciumKing committed:
Creating Login Form Functionality

- In login.ui: made a clean form ui to login using a username, no password is required because nothing is required for the current system.

- In main.ui: changing instances of "exit" to "Log Out"

- In scanner.py: simplified some logic, added form functionality, and some error handling. Added logs for users logging out.

- In app.py: changed the exit button into the login button and made it log out, traversing to login.ui instead of exiting.

- In abstract.py: made `to_page()` more efficient using super and its @override method

CalciumKing committed:
Images, Icons, Styles, Big .ui Changes

- Added lots of icons to be used around application, no functionality changes, just visuals.

- "Make the sidebar buttons look like buttons" per Jasmines request

- In .ui files: added the icons in appropriate places.

- In export.ui: changed texts to say "Export To:" and "In Folder:".

- In resource.qrc: renamed it to resources.qrc and added data into it. It currently does nothing because my laptop cant convert the file into a _rc.py file, that will happen at the end of the project.

CalciumKing committed:
QR Generate Page, Merged Export Classes

- Created qr_generate.ui and is controlled by QRGenerate() in export.py.

- Merged qr_generator.py and file_exports.py Into ExportUtils().

- In abstract.py: fixed some docstrings, created AbstractExporter() to control some common logic and variables in Export and QRGenerate classes, both in export.py

- In app.py: added QR Generate page to list of controllers.

- In main.ui: added a QR Generate button and changed some icons around.

- In enums.py: added a GENERATE enum to control the generator page.

- In file_exports.py: changed class from FileExports to ExportUtils, fixed problems with get_valid_name, optimized instances of AbstractController and App

- Added test_qr_generation.py to test the original feature.

CalciumKing committed:
Fixed export issue

- In export.py: changed case variables to the values and not the actual enum itself, removed self from the method call in csv/tsv/psv which was causing issues and wasn't needed.

- In export.ui: removed the progress bar and added a vertical spacer to make the page look better.

- In file_exports.py: changed enum parameter from enum to string to work with match-case statement, removed spaces from psv delimiter to prevent formatting bug in .psv files. In exception, made a local uppercase variable to use in error messages instead of calling .upper() several times.

CalciumKing committed:
Made page assigning automatic with enums, working towards fully automatic page creation.

- By assigning a class to an enum in enums.py, a variable is created in app.py to automatically assign new pages so the only place it needs to be entered is in enums.py and main.ui.

- In app.py: removed lots of controller initialization and replaced with an automatic creation based on the page enums in enums.py.

- In enums.py: changed TITLE to PAGE_TITLE, moved GENERATE page, and added class CONTROLLER attributes to make it easier to create a new page.

CalciumKing committed:
Moved Title Setting From app.py to abstract.py

- Deleted page title changing from app.py because it makes more sense to have it in to_page()

- Added the try-except to abstract.py, only changing a few details for efficiency.

CalciumKing committed:
Added application shortcuts and some optimization

- In app.py: added shortcuts located in main.ui to be handled in handle_connections(). Moved handle_connections() out of __init__ because I don't want a 76 line __init__ method. Fixed a bug where log out button wouldn't work.

- Added shortcuts and corresponding methods for the following:
  - Page selection (1-Page #)
  - Toggle Maximize (F)
  - Escape Maximize (ESC)
  - Minimize (M)
  - Search (CTRL+F)
  - Log Out (CTRL+Q)
  - Close Application (CTRL+X)

- Updated and added some docstrings

- In *.ui: made line edits not auto focusable so the user can freely flip between pages without getting stopped and made text edits tab traversable to go to the next field.

- In login.ui: fixed no button styles error by removing active css. Made username field returnable so the user can just press enter instead of pressing login button.

- In main.ui: added a menu bar to allow shortcuts and manual selection if needed.

- Updated resources.qrc

- Added .svg images for all the new actions, so they have icons in the shortcuts menu bar.

CalciumKing committed:
scanner.ui color update, README.md/meta.yaml updates

- Slightly changed the layout and functionality of scanner.ui to be more visually appealing and prevent altering of data with the QTextEdit.

- Updated README.md with new and improved descriptions, file tree, requirements, git information, etc.

- Updated meta.yaml and requirements.txt with qrcode build requirement.

CalciumKing committed:
Added docstrings to recently created methods

tangkong commented on assets/icons/add.svg:
Rather than including all the icons here, we usually use [`fontawesome`](https://pypi.org/project/fontawesome/) to provide icons.  This is installed in our system already, so the cost of adding it to your dependencies is low

CalciumKing replied:
I got the icons from the fontawesome.com, I didn't know that they had a python package, Ill look into this.

tangkong commented on stock_manager/app.py:
```diff
@@ -32,31 +31,30 @@ def __init__(self):
-		self.user: str | None = None
+		self.export_utils = ExportUtils(self)
+		
+		self.controllers: dict[str, type[stock_manager.AbstractController]] = {
```
Are you sure your dictionary here is taking the type-class `AbstractController`?  Or are its values actually instances of `AbstractController`?

CalciumKing replied:
Each of the `AbstractController` objects in that dictionary are types as stated in `enums.py` (e.g. `stock_manager.Login` and not `stock_manager.Login()`). I do this because each of them need a reference to `App()` and only a few lines later I initialize them with an instance of `App()` to make the process of creating a page instance dynamic so its easier to create pages.

tangkong commented on stock_manager/app.py:
```diff
@@ -69,31 +67,38 @@ def __init__(self):
+		"""
+		Connects sidebar buttons to the appropriate screen
+		navigation actions and connects application shortcuts
+		to a method.
```
Which method? Can you be more descriptive here?  (since you're already going to the effort of adding a docstring)

tangkong commented on stock_manager/app.py:
```diff
@@ -69,31 +67,38 @@ def __init__(self):
-				self.screens.addWidget(screen)
-			
-			self.screens.currentChanged.connect(self._on_page_changed)
+		self.buttons: list[QPushButton] = [child for child in self.sideUI.children()]
```
While it may seem efficient to store all the buttons in one place, it's not very readable.  How do I know what buttons are in what index?  How do I know that buttons will end up in the same order?  (.children() retrieves objects in the order their added, but are you the one adding them or is the ui file doing that?)

Generally it's better to explicitly name objects instead of relying on an implicit indexing that future developers will have to decipher.

CalciumKing replied:
I do this and the `self.controllers` dictionary so that when creating a new page, the user only has to create a new enum in `enums.py`. This is what I thought you meant by making everything "automatic" instead of changing multiple values for one new addition. Should I get rid of the currently dynamic system of controllers and buttons and replace it with assigned values?

e.g.:
Replace currently lines 42-49 with:
```
self.login = Login(self)
self.view = View(self)
self.item_scanner = ItemScanner(self)
...
```

tangkong replied:
This is sort of what I meant.  I do appreciate you centralizing the page numbers, and think the enum solution is a good one for the job.  I think you may have just taken one step farther than what I was thinking

My specific issue here is: if I want to grab the "add" button in `app.py`, I have to know a specific index.  

I think they way you had it before was more explicit and readable, while also compatible with the centralized page number assignment you set up in enums.py, right?

tangkong replied:
Ah I see now that you're recreating the previous set of attributes, just dynamically.  I apologize, I missed that on my first and second readthroughs.

I do think listing the attributes explicitly is more readable, helps enable our static type checkers, and doesn't cost us too much.  (we're not optimizing on LoC here).  But I'll leave the final decision here up to you

CalciumKing replied:
I agree, I think that static type checkers and readability is worth the extra few steps of adding a reference in `App().__init__()`, I will change some of the code back to what it was before.

tangkong commented on stock_manager/app.py:
```diff
@@ -69,31 +67,38 @@ def __init__(self):
+		button: QPushButton
+		controller: stock_manager.AbstractController
+		enum: stock_manager.Pages
+		for button, controller, enum in zip(self.buttons[:-1], self.controllers.values(), stock_manager.Pages):
```
Related to the above comment on arbitrarily indexed lists, what guarantee do I have that the last button is the same every time?  If it's going to be left out here, why is it in the list?

If we were to simply create a list the buttons to iterate here by name, we'd at least be able to verify at a glance that we're modifying the right objects.

CalciumKing replied:
Along with the fix of the above comment, I fixed this to not include magic number `[:-1]` and made `self.buttons` a list of manually assigned values. Slightly more effort than what system I had, but much more readable.

tangkong replied:
Thanks!  It's a balance we have to strike.  We should try to be efficient, but at the same time shouldn't play code golf.

tangkong commented on stock_manager/controllers/add.py:
```diff
@@ -176,7 +176,7 @@ def empty_fields_check() -> bool:
 		self.app.all_items.append(new_item)
-		self.logger.info_log(f'Item Added To Database: {new_item.part_num}')
+		self.logger.info_log(f'{self.app.user} Added Item To Database: {new_item.part_num}')
 		self.app.update_tables()
```
your update_tables() method is a coroutine, calling it without an `await` simply creates the coroutine and does not execute its contents.  

This is true for the other places you call `update_tables`

CalciumKing replied:
I have it as an `async def` so that I am able to call `await` in it. I thought that the purpose of `@asyncSlot()` decorator was to load the contents of the function asynchronously, is that wrong? The method does work when called because all the tables do get updated properly.

tangkong replied:
I did see those asyncSlot decorators, but not in proximity to this method.  For example you await [here](https://github.com/CalciumKing/SLAC-LCLS-Stock-Management/blob/01b45088aa0e1f8c00c17b1c7c12b1cf8d0af8c9/stock_manager/app.py#L144), but not in this file.  

Is there an await in this method call specifically?  Am I missing something?

CalciumKing replied:
It is to my understanding that if a method has an `@asyncSlot()` decorator, it does not need to be awaited but can still be asynchronous, I found that adding that decorator is the equivalent of writing `await asyncio.create_task(method_name())`. I do not call `.update_tables()` in any other `async` methods, and since this method is already `async` because of it's `@asyncSlot()` decorator, I figured it is good practice to await it even through technically it is not required.

tangkong commented on stock_manager/controllers/abstract.py:
```diff
@@ -167,9 +178,6 @@ def __init__(self, file_name: str, app: 'App'):
-		
 		try:
 			self.start_video()
 		except Exception as e:
```
It's bad form to catch generic exceptions.  You should be specific about what exception you want to catch, and only catch that exception. 

If your code throws an error you don't expect, it may still be caught by this block and considered "ok".

CalciumKing replied:
I understand this and where I can I do catch specific exceptions, however with python packages that I use in this project, it doesn't tell me what errors could arise or even that there could be errors at all. With languages that I have experience with like Java, anything that can throw an error either directly tell you in the documentation or displays in the IDE. It seems that isn't the case with python, and I am unaware of how to find out what errors something can raise. I think my python error handling skills could use some work, I feel my error handling throughout this entire project is pretty basic and lacking.

tangkong replied:
You're right that normally it should be easier to find out what specific errors might get thrown, and well written python should also do this.  If you're having trouble finding an appropriate exception to catch here, the code probably doesn't expect to raise errors (or the library isn't thorough)

At this point I think it's less a problem with your skills and more a problem with whatever dependencies you're using.  In this case I saw we were catching an error from one of our methods, and figured we were working around a known error.

CalciumKing replied:
Not exactly, I wrapped `.start_video()` in a `try-except` because if an error occurs in that method it would be logged properly. However, something that I cant seem to figure out is whether python has a sort of "error chaining" that Java has. In Java if a method throws an error, even if the error is handled, it is still passed up through all the methods, being logged in multiple places for the same singular issue, that is why I very commonly wrap method calls in error handlers even if there are already error handlers in those methods. But since this method is only being called once and everything in the method is already wrapped in a `try-except` I'm not sure if the `try-except` in `.to_page()` is necessary. If not, I could also look throughout my program and remove any unnecessary error handling that I may have in place.

tangkong replied:
I see where you're coming from.  To the best of my understanding / memory, exception handling is similar between java and python.  In python, if an exception is thrown and not caught, it will print a stack trace and end your program.  If it's caught, you get to choose what to do do next, and the error propagation ends there. 

I would argue that unless you expect the app to throw an error during **normal** operation, you let the error propagate and end the program.  If you do expect it, you handle it appropriately for that context.  Otherwise the app can break silently

tangkong commented on stock_manager/controllers/export.py:
```diff
@@ -30,35 +32,25 @@ def __init__(self, app: 'App'):
 		super().__init__(page.value.FILE_NAME, app)
 		self.PAGE_NAME = page
-		self._path = str(Path(__file__).resolve().parent.parent.parent / 'exports')
+		self.location_btn.setText(f'...{self.path[-6:]}')
```
Magic number warning here.  You've picked an index here that will break if we change filenames.  Try splitting the filepath on relevant delimiters instead.

tangkong commented on stock_manager/controllers/export.py:
```diff
@@ -86,33 +78,75 @@ def _export_data(self) -> None:
+		super().__init__(page.value.FILE_NAME, app)
+		self.PAGE_NAME = page
+		self._selected_qr: BaseImage | None = None
+		self.location_btn.setText(f'...{self.path[-6:]}')
```
Similar magic number problem to above

tangkong commented on stock_manager/controllers/view.py:
```diff
@@ -35,4 +35,4 @@ def __init__(self, app: 'App'):
 	def handle_connections(self) -> None:
 		self.search.textChanged.connect(self.filter_table)
-		self.export_btn.clicked.connect(self.app.export.to_page)
+		self.export_btn.clicked.connect(lambda: self.app.export.to_page())
```
Why did you need to wrap this in a lambda?  Was your callback getting fed arguments it didn't expect?  All signals call their callbacks with some additional arguments, and normally you should write your callbacks to properly handle those arguments.

Further down the rabbit hole, assigning anonymous functions to slots can be a bit dangerous.  I think in this case we'll be ok, but in some cases the QT gargage collector can bite us.

CalciumKing replied:
I assign `AbstractController`'s `to_page()` method to a `lambda` to defer its execution until runtime. Without the `lambda`, the method is called immediately, and `app.export` may not exist yet.

When I do not use `lambda`, I get this error:
`AttributeError: 'App' object has no attribute 'export'`

By wrapping it in a `lambda`, I ensure that the method call is delayed until the user triggers the event, by which time all instances have been created. I do this in a few other places for the same reason.

tangkong replied:
I see.  That's a bit unfortunate, but an interesting workaround.  

I'd have probably written a method that checks for the existence of app.export, and called `.to_page()` if it does, but I guess this achieves the same result

tangkong reviewed (commented):
Some good changes here.  I left a few comments, but I wanted reiterate the comments I left in the last PR that have not been addressed.  

* We use qt5 at LCLS. It'd be even better here if we used the qtpy package, which is a translation layer that makes our code agnostic to the precise qt binding. I've seen the qt5->qt6 migratio effort be quite painful, so it's probably best to make this change early.
* This needs a general style pass. We abide by [pep8](https://peps.python.org/pep-0008/) standards broadly, e.g. tabs -> spaces (4 spaces)

The first of these is an actual problem that will prevent us from using this application.  qt6 is the most recent Qt, but we don't support it here yet, and the migration is non-trivial.

The second is a lesson to take to whatever future software development teams you join in the future.  Having standard style helps keep code readable and maintainable by everyone in the team.  Python in particular has a well defined set of style standards that we use here at ECS, and it's important to abide by them.  It may seem like a petty thing, but code is meant for humans to read first and foremost.  

If you'd like help making your codebase compliant, we have tools like [pre_commit](https://github.com/pcdshub/cookiecutter-pcds-python/blob/master/%7B%7B%20cookiecutter.folder_name%20%7D%7D/.pre-commit-config.yaml) to help automate these things.  The sooner you start following the standards, the less pain you'll go through trying to fix up your code in the future

CalciumKing committed:
Requested PR#2 Changes - Commit 1

- Added specificity to handle_connections() in app.py.

- Added docstrings for everything in QRGenerate() class in export.py.

- Removed magic numbers from location_btn in abstract.py and export.py.

CalciumKing committed:
Requested PR#2 Changes - Commit 2

- Changed all tab whitespace in all .py files to 4 spaces

- This commit is ONLY whitespace changes

CalciumKing committed:
Requested PR#2 Changes - Commit 3

- Changed all instances of PyQt6 to PyQt5

- Optimized some QMessageBoxes because PyQt5 is slightly different from PyQt6. Replaced all instances of QMessageBox.StandardButton... with just QMessageBox...

- In abstract.py: removed all QMessageBoxes in _CameraThread because it causes a crash, this didn't happen with PyQt6.

CalciumKing committed:
Auto reformatting of .ui icons

- When saving .ui files, the icon sections of the code get reduced. Went through all files and saved to reduce line count

tangkong reviewed (commented):
Thanks for the responses!  I think I understand a little better now

ZLLentz commented:
Robert was really thorough here- I didn't have anything else to add. Forward progress screenshots look good, hopefully you've been getting good feedback.

CalciumKing committed:
Requested PR #2 Changes - Commit 4

- In app.py: removed dynamic but confusing page and button system for manually assigned values that are significantly easier to understand and read. Due to this change, other lines had to change too, but overall better readability changes. Added return type annotation for shortcut methods.

- in enums.py: Removed CONTROLLER from _PageDetails(). Removed controller type parameter from all the class references in this file. Removed all references to enum controllers in the only other file its referenced in (app.py).

CalciumKing committed:
Requested PR #2 Changes - Commit 5

- Removed all absolute file icons from project and .ui files. Assigned the icons dynamically in each controller.

- New dynamic icons are assigned in .handle_connections(), but some controllers with more icons and settings are located in .handle_connections().handle_icons().

- Fixed some problems with search boxes not appearing as they do in editor.

- Updated README.md, meta.yaml, and requirements.txt with qtawesome requirement.

CalciumKing committed:
Requested PR #2 - Commit 6

- Changed all QTableWidgets to QTableViews, making the process of selecting items, customization, and searching for items more efficient.

- Changed all references of QTableWidgets to QTableViews

- Changed all ._on_cell_clicked() parameters to use a QModelIndex instead of a row and a column.

- Moved self.search.textChange... to .update_table() in abstract.py since all the search bars do the same thing and are necessary for a table.

- Deleted .filter_table() from abstract.py since its less efficent than the primary purpose of .setFilterFixedString().

- Updated some related docstrings.

- Added .get_headers() to database.py in order to set the correct table headers for each table and be responsive with the GS database.

CalciumKing committed:
Requested PR #2 Changes - Commit 7

- Added standard.yml, only changed 'package-name', not sure if there is anything else I need to change.

- Removed icon files from resources.qrc.

- Changed constant handing down of app.logger to more robust logging.getLogger() and its own logging methods like info, warning, and error instead of info_log, warning_log, and error_log.

CalciumKing committed:
Requested PR #2 Changes - Commit 8

- Changed 'test' folder to 'scripts' because it only tests standalone features and has nothing to do with unit testing.

- Created the basics of a 'tests' folder.

- This is a strictly organizational commit

CalciumKing committed:
Requested PR #2 Changes - Commit 9

- Replaced magic string for google sheet file name with a constants.py value, updated README.md to explain how to update that constant with the sheet name when starting project.

- Added instances of GS_FILE_NAME where needed for this to work.

- Also included some error messages that were missed in the last commit.

CalciumKing commented:
I believe my most recent commits and push should have resolved a lot of the feedback given for this to go forward. I think we can safely merge and I can continue to work on unit testing and possibly expanding the project after a meeting I have tomorrow.

tangkong reviewed (approved):
We can merge and move on.  👍
